### ISSUE TITLE
Solution to check if a requirement is met without pkg_resources

### ISSUE THREAD

**Issue by HansBug** [pypa/packaging-problems#664]
### Problem description

In the previous `pkg_resources` package, this issue can be solved with the following code ([link](https://stackoverflow.com/questions/16294819/check-if-my-python-has-all-required-packages/16298328#16298328)):

```python
import pkg_resources
from pkg_resources import DistributionNotFound, VersionConflict

# dependencies can be any iterable with strings, 
# e.g. file line-by-line iterator
dependencies = [
  'Werkzeug>=0.6.1',
  'Flask>=0.9',
]

# here, if a dependency is not met, a DistributionNotFound or VersionConflict
# exception is thrown. 
pkg_resources.require(dependencies)
```

However, `pkg_resources` has been deprecated, so this method is not recommended for continued use. In the [migration guide](https://importlib-metadata.readthedocs.io/en/latest/migration.html#pkg-resources-require) provided in the `importlib_metadata` documentation, there is no new interface that is completely equivalent to `pkg_resources.require`. Therefore, we need to use the `import_metadata` library (native in Python 3.8 or higher) and the `packaging` library to achieve this functionality.

I have currently implemented a version, the code is here: https://github.com/HansBug/hbutils/blob/main/hbutils/system/python/package.py#L202, where the most critical part is the `_yield_reqs_to_install` function, which checks a requirement and its sub-requirements included in its extra one by one, and enumerates the unsatisfied requirements. By calling this iterator function, the check of whether a requirement is satisfied can be achieved. This is the core code of the function:

```python
def _yield_reqs_to_install(req: Requirement, current_extra: str = ''):
    if req.marker and not req.marker.evaluate({'extra': current_extra}):
        return

    try:
        version = importlib_metadata.distribution(req.name).version
    except importlib_metadata.PackageNotFoundError:  # req not installed
        yield req
    else:
        if req.specifier.contains(version):
            for child_req in (importlib_metadata.metadata(req.name).get_all('Requires-Dist') or []):
                child_req_obj = Requirement(child_req)

                need_check, ext = False, None
                for extra in req.extras:
                    if child_req_obj.marker and child_req_obj.marker.evaluate({'extra': extra}):
                        need_check = True
                        ext = extra
                        break

                if need_check:  # check for extra reqs
                    yield from _yield_reqs_to_install(child_req_obj, ext)

        else:  # main version not match
            yield req

def check_reqs(reqs: List[str]) -> bool:
    """
    Overview:
        Check if the given requirements are all satisfied.

    :param reqs: List of requirements.
    :return satisfied: All the requirements in ``reqs`` satisfied or not.

    Examples::
        >>> from hbutils.system import check_reqs
        >>> check_reqs(['pip>=20.0'])
        True
        >>> check_reqs(['pip~=19.2'])
        False
        >>> check_reqs(['pip>=20.0', 'setuptools>=50.0'])
        True

    .. note::
        If a requirement's marker is not satisfied in this environment,
        **it will be ignored** instead of return ``False``.
    """
    return all(map(lambda x: _check_req(Requirement(x)), reqs))
```

Of course, if you need this function immediately, you can simply 

```shell
pip install hbutils>=0.9.0
```

and then

```python
from hbutils.system import check_reqs
print(check_reqs(['pip>=20.0']))
print(check_reqs(['pip~=19.2']))
print(check_reqs(['pip>=20.0', 'setuptools>=50.0']))
```

`hbutils` is a universal toolkit library containing various common functions, and the project address is https://github.com/HansBug/hbutils. It will be maintained for a long time.

**Comment by HansBug** [comment]
From https://github.com/python/importlib_metadata/issues/450

**Comment by smheidrich** [comment]
@HansBug Hi! I'm currently trying to get some information about how to best detect & handle missing extras / optional dependencies at runtime into the PyPA's packaging guide (https://github.com/pypa/packaging.python.org/pull/1606) and your `check_reqs` is the only modern alternative to `pkg_resources.require` I know of.

The [current draft](https://python-packaging-user-guide--1606.org.readthedocs.build/en/1606/guides/handling-missing-extras-at-runtime/) links to both this issue and your `hbutils` package, but it feels a bit weird to recommend a "giant toolbox of stuff"-type package for this one function. I'm wondering if you'd be open in principle to splitting it off into its own package? If you don't want to do it yourself, I could also do it & make you maintainer afterwards.

Not saying that it should be done _now_, because it's still possible that PyPA's reviewers will decide the guide shouldn't include this information at all, in which case it'd be pointless effort - but I thought I'd ask here & give a heads-up before making promises over there.

**Comment by sinoroc** [comment]
Maybe [_distlib_](https://pypi.org/project/distlib/) can do this. Within the Python packaging ecosystem, it is a relatively well-known and definitely well-trusted library. If it can not do it out of the box, maybe the amount of custom code necessary would be smaller than what would be necessary when using the combination of `importlib.metadata` and `packaging`. I do not know for sure, but maybe it is worth investigating.

**Comment by smheidrich** [comment]
I had a look at `distlib` and it doesn't seem like it has anything to make this more comfortable than with `importlib.metadata` and `packaging`. I thought maybe something like this might work:

```python
from distlib.locators import DistPathLocator, DependencyFinder
from distlib.database import DistributionPath

req = "typer-slim[standard]"
distpath = DistributionPath(include_egg=True)
locator = DistPathLocator(distpath)
print(locator.locate(req))
```

But unfortunately it seems to completely ignore the extra in the requirement and always returns a positive result whether it's installed or not.

I then tried

```python
dfin = DependencyFinder(locator)
print(dfin.find(req))
```

But this goes in the other direction and reports anything as unsatisfied that's part of _any_ extra, whether the extra is part of the requirement or not, plus some other weird stuff like dependencies whose marker is not even fulfilled...

**Comment by sinoroc** [comment]
I asked: <https://github.com/pypa/distlib/issues/234>.

**Comment by sinoroc** [comment]
Cross link:

https://discuss.python.org/t/check-if-an-environments-packages-conflict/67511

**Comment by smheidrich** [comment]
By the way, another reason why even _if_ this functionality was in `distlib`, a separate package might still be preferable: distlib's minimum [download size](https://pypi.org/project/distlib/#files) is around 500 kB and its installed size almost 2 MB. In many cases that will be larger than the dependencies the package author is trying to make optional, defeating the purpose. Contrast this with the [packaging package](https://pypi.org/project/packaging/#files) whose wheel is only 50 kB.

**Comment by sinoroc** [comment]
> distlib's minimum [download size](https://pypi.org/project/distlib/#files) is around 500 kB and its installed size almost 2 MB. In many cases that will be about as large or larger than the dependencies the package author is trying to make optional, defeating the purpose

Good point

**Comment by smheidrich** [comment]
@HansBug Another question: It's probably implied by you posting this snippet here for others to use, but just to be sure: Is it fine to include it (or modified versions of it) in the [aforementioned guide](https://github.com/pypa/packaging.python.org/pull/1606) verbatim, for people to copy into their code, _without_ the restrictions imposed by the Apache 2 license under which `hbutils` is provided?

**Comment by vsajip** [comment]
> distlib's minimum download size is around 500 kB and its installed size almost 2 MB

Yes, and that includes the launchers which are used on Windows only - those are around 750K in size.

I can look at the issue @sinoroc raised on distlib, but if the size is going to be a deal-breaker for your needs, might you not use it, anyway?

**Comment by smheidrich** [comment]
It would depend on the use case, e.g. if the optional dependency in question is NumPy or PyTorch, pulling in a 500 kB dependency should be worth it to make the experience better.

But just going by the description of distlib's purpose from its README stating

> It is intended to be used as the basis for third-party packaging tools

it doesn't sound like the natural place for such functionality, which would be meant to be used in arbitrary packages, not just packaging tools.

**Comment by vsajip** [comment]
I posted a [possible solution](https://github.com/pypa/distlib/issues/234#issuecomment-2414807376) on the distlib issue.
